### PR TITLE
TRM-32125: Improve RequestConverter for ids endpoints

### PR DESCRIPTION
**Pull request by LeonardoGonzales**

### CODE REVIEW

File: common-rest/src/main/java/org/uniprot/api/rest/service/request/BasicRequestConverter.java [hunk 1]
```diff
@@ -62,6 +70,32 @@ public SolrRequest.SolrRequestBuilder createStreamSolrRequest(StreamRequest requ
...
+    private void updateIdsSort(SolrRequest.SolrRequestBuilder requestBuilder, String requestSorts) {
+        requestBuilder.clearSorts();
+        if (Utils.notNullNotEmpty(requestSorts)) {
+            requestBuilder.sorts(
+                    solrSortClause.getSort(requestSorts).stream()
+                            .filter(clause -> !clause.getItem().equals("score"))
+                            .toList());
+        }
+    }
+
```
Comment (ahmadshadab) [inline common-rest/src/main/java/org/uniprot/api/rest/service/request/BasicRequestConverter.java:98]: We can extract the code inside requestBuilder.sorts() to a method. And lets reuse SCORE const of AbstractSolrSortClause.
Probably, we can use equals like this !"score".equals(clause.getItem())

File: common-rest/src/main/java/org/uniprot/api/rest/service/request/BasicRequestConverter.java [hunk 2]
```diff
@@ -76,24 +110,43 @@ public String getQueryFields(String query) {
...
-                requestedQuery = cleanQuery.toUpperCase();
+        String query = request.getQuery();
+        if (query != null) {
+            if (idRequestRegex != null) {
+                String cleanQuery =
+                        CLEAN_QUERY_REGEX.matcher(request.getQuery().strip()).replaceAll("");
+                if (idRequestRegex.matcher(cleanQuery.toUpperCase()).matches()) {
+                    query = cleanQuery.toUpperCase();
+                }
             }
```
Comment (ahmadshadab) [inline common-rest/src/main/java/org/uniprot/api/rest/service/request/BasicRequestConverter.java:143]: We should explain the logic here for new dev with an example p12345

File: common-rest/src/test/java/org/uniprot/api/rest/service/request/BasicRequestConverterTest.java [hunk 3]
```diff
@@ -285,4 +308,66 @@ void createStreamBasicSolrRequest() {
...
+        Mockito.when(request.getQuery()).thenReturn(query);
+        Mockito.when(request.getSort()).thenReturn(sort.toString());
+
+        SolrRequest result = converter.createIdsSolrRequest(request, idField).build();
+        assertNotNull(result);
+        assertEquals(query, result.getQuery());
+        assertEquals(idField, result.getIdField());
+        assertEquals(List.of(sort), result.getSorts());
+    }
 }
```
Comment (ahmadshadab) [inline common-rest/src/test/java/org/uniprot/api/rest/service/request/BasicRequestConverterTest.java:373]: Add test for method getIdsTermQuery

File: common-rest/src/main/java/org/uniprot/api/common/repository/search/facet/FacetConverter.java [hunk 4]
```diff
@@ -1,14 +1,14 @@
...
+import java.util.Collection;
 import java.util.Map;
 import java.util.Optional;
 
 import org.uniprot.core.util.Utils;
 
 public abstract class FacetConverter<F, T> {
 
-    public abstract T convert(F queryResponse, List<String> facetList);
+    public abstract T convert(F queryResponse, Collection<String> facetList);
```
Comment (ahmadshadab) [inline common-rest/src/main/java/org/uniprot/api/common/repository/search/facet/FacetConverter.java:12]: I couldn't understand the need to change the 2nd parameter from List to Collection.

File: common-rest/src/main/java/org/uniprot/api/common/repository/solrstream/FacetStreamExpression.java [hunk 5]
```diff
@@ -28,99 +28,54 @@ public enum MetricFunctionName {
...
-                        "bucketSorts", getBucketSorts(request, facetProperty)));
+                new StreamExpressionNamedParameter("bucketSorts", getBucketSorts(facetRequest)));
         this.addParameter(
                 new StreamExpressionNamedParameter(
-                        "bucketSizeLimit", String.valueOf(request.getBucketSizeLimit())));
-        List<StreamExpression> metricExpressions = parseMetrics(request.getMetrics());
-        this.getParameters().addAll(metricExpressions);
+                        "bucketSizeLimit", getBucketSizeLimit(facetRequest)));
+        StreamExpression expression =
+                new StreamExpression(MetricFunctionName.count.name()).withParameter("*");
```
Comment (ahmadshadab) [inline common-rest/src/main/java/org/uniprot/api/common/repository/solrstream/FacetStreamExpression.java:44]: Earlier we were getting list of StreamExpression for metrics. Now, we have hardcoded only one metric count(*). Do we only need one metric?

File: idmapping-common/src/main/java/org/uniprot/api/idmapping/common/service/BasicIdService.java [hunk 6]
```diff
@@ -8,27 +8,22 @@
 import org.apache.solr.client.solrj.io.stream.TupleStream;
```
Comment (ahmadshadab) [inline idmapping-common/src/main/java/org/uniprot/api/idmapping/common/service/BasicIdService.java:8]: I think, it is better if we could deploy this branch in dev and do idmapping testing through the uniprot website.
1. Few Idmapping jobs
2. IdMapping with facets and sort
3. just ids etc

File: common-rest/src/main/java/org/uniprot/api/common/repository/solrstream/SearchStreamExpression.java [hunk 7]
```diff
@@ -13,37 +14,37 @@
...
+            throw new IllegalArgumentException("query or ids is a mandatory param");
         }
-        if (Utils.nullOrEmpty(request.getSearchFieldList())) {
+        if (Utils.nullOrEmpty(request.getQueryField()) && Utils.nullOrEmpty(request.getIdField())) {
             throw new IllegalArgumentException("fl is a mandatory param");
         }
-        if (Utils.nullOrEmpty(request.getSearchSort())) {
+        if (Utils.nullOrEmpty(request.getSorts()) && Utils.nullOrEmpty(request.getIdField())) {
             throw new IllegalArgumentException("sort is a mandatory param");
         }
```
[the remark this answers is not in the document]
Comment (LeonardoGonzales) [reply common-rest/src/main/java/org/uniprot/api/common/repository/solrstream/SearchStreamExpression.java:48]: We uses idField in sort logic

File: common-rest/src/main/java/org/uniprot/api/common/repository/solrstream/FacetStreamExpression.java (hunk 5 above)
Comment (LeonardoGonzales) [reply common-rest/src/main/java/org/uniprot/api/common/repository/solrstream/FacetStreamExpression.java:44]: Yes, we currently only uses it.

**Review by ahmadshadab — APPROVED** [review]

**Pull request merged**